File: to-lower-case/Solution709.java
// https://leetcode.com/problems/to-lower-case/

import java.util.HashMap;
import java.util.Map;

class Solution709 {

    private static final int DIFF = 'a' - 'A'; // 32
    private static final String UPPER = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final String LOWER = "abcdefghijklmnopqrstuvwxyz";
    private static final Map<Character, Character> MAP = new HashMap<>();
    static {
        for (int i = 0; i < 26; ++i) {
            MAP.put(UPPER.charAt(i), LOWER.charAt(i));
        }
    }

    // T: O(n), S: O(n)
    public String toLowerCase(String s) {
        var sb = new StringBuilder();
        for (char ch : s.toCharArray()) {
            if (ch >= 'A' && ch <= 'Z') {
                sb.append((char) (ch + DIFF));
            } else {
                sb.append(ch);
            }
        }
        return sb.toString();
    }

    // T: O(n), S: O(n)
    public String toLowerCaseMap(String s) {
        var sb = new StringBuilder();
        for (char ch : s.toCharArray()) {
            sb.append(MAP.containsKey(ch) ? MAP.get(ch) : ch);
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        var sol = new Solution709();
        var s1 = "Hello";
        var s2 = "here";
        var s3 = "LOVELY";
        System.out.println(sol.toLowerCase(s1)); // hello
        System.out.println(sol.toLowerCaseMap(s1)); // hello
        System.out.println(sol.toLowerCase(s2)); // here
        System.out.println(sol.toLowerCaseMap(s2)); // here
        System.out.println(sol.toLowerCase(s3)); // lovely
        System.out.println(sol.toLowerCaseMap(s3)); // lovely
    }

}

File: to-lower-case/solution709.py
# https://leetcode.com/problems/to-lower-case/

DIFF = ord('a') - ord('A')  # 32
UPPER = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
LOWER = "abcdefghijklmnopqrstuvwxyz"
MAP = {UPPER[i]: LOWER[i] for i in range(26)}


class Solution709:

    # T: O(n), S: O(n)
    def toLowerCase(self, s: str) -> str:
        result = []
        for ch in s:
            if 'A' <= ch <= 'Z':
                result.append(chr(ord(ch) + DIFF))
            else:
                result.append(ch)
        return ''.join(result)

    # T: O(n), S: O(n)
    def toLowerCaseMap(self, s: str) -> str:
        result = []
        for ch in s:
            result.append(MAP.get(ch, ch))
        return ''.join(result)


def main() -> None:
    sol = Solution709()
    s1 = "Hello"
    s2 = "here"
    s3 = "LOVELY"
    print(sol.toLowerCase(s1))      # hello
    print(sol.toLowerCaseMap(s1))  # hello
    print(sol.toLowerCase(s2))      # here
    print(sol.toLowerCaseMap(s2))  # here
    print(sol.toLowerCase(s3))      # lovely
    print(sol.toLowerCaseMap(s3))  # lovely


if __name__ == "__main__":
    main()
